### PR TITLE
Support passing a context to BusOject Call and Go

### PR DESCRIPTION
**Pull request by dannin**
Sometimes a method call may take too much time replying or may never reply, causing message to block indefinitely.  Support adding a context.Context which can help.

### CODE REVIEW

**Comment by dannin** [comment]
@jsouthworth Are you able to take a look at this? Thanks

**Comment by jsouthworth** [comment]
Sorry for the delay. It'll take me a bit longer to review this. I am unfamiliar with the context package.

File: conn.go [hunk 1]
```diff
@@ -386,12 +400,19 @@ func (conn *Conn) Send(msg *Message, ch chan *Call) *Call {
 		call.Method = iface + "." + member
 		call.Args = msg.Body
 		call.Done = ch
+		call.ctx = ctx
+		call.ctxCanceler = canceler
 		conn.calls.track(msg.serial, call)
+		go func() {
+			<-ctx.Done()
+			conn.calls.handleSendError(msg, ErrClosed)
```
Comment (jsouthworth) [inline conn.go]: Do lines 408 and 411 cause duplicate calls to handleSendError or does that not matter the way this is used?

File: conn.go [hunk 2]
```diff
@@ -736,32 +752,68 @@ func (tracker *callTracker) handleSendError(msg *Message, err error) {
...
 }
 
 func (tracker *callTracker) finalizeAllWithError(err error) {
-	closedCalls := make(map[uint32]*Call)
-	tracker.lck.RLock()
-	for sn, v := range tracker.calls {
-		v.Err = err
-		v.Done <- v
-		closedCalls[sn] = v
+	var closedCalls []*Call
```
Comment (jsouthworth) [inline conn.go]: You can preallocate the array to the size of the map to avoid resizing. 
```
closedCalls := make([]*Call, 0, len(tracker.calls))
```

**Comment by dannin** [comment]
Opened a new pull request

**Pull request closed without merging**